### ISSUE TITLE
Normalized cache does not try to normalize internal objects if query caching fails

### ISSUE THREAD

**Issue by bpollock-vida** [apollographql/apollo-android#1136]
I have found the Android Apollo client to behave differently than the iOS client. Making the normalized caching unusable.

On iOS I use the cacheKeyForObject to set a cache key. This method is evaluated once for the entire response and once for each object in the response. For example, if returned { "object": [Object] }. It would be evaluated for the dictionary containing "object" and also for each Object. Furthermore, if there was a cache miss, the apollo client would still return a result.

Code:
```
client.cacheKeyForObject = {
return $0["urn"]
}

However, on Android. The resolver only evaluates once for the entire response, making it impossible to do any cache normalization. Secondly, the graphql query doesn't seem to emit a response when there is a cache miss.

val query = GetTagsForGroupQuery(group, size, start)

return Rx2Apollo.from(apolloClient.query(query))
.map { getTagsResult -> Result.fromResponseRequiredData(getTagsResult, LOG_TAG) }
// this map function is not evaluated upon cache miss.

Bugs:
1. resolver does not evaluate each object in response, just top level
2. on cache miss, the query does not emit values
3. There is no SqlNormalizedCacheFactory as mentioned in documentation

Code:
CacheKeyResolver resolver = new CacheKeyResolver() {
@NotNull @Override
public CacheKey fromFieldRecordSet(@NotNull ResponseField field, @NotNull Map<String, Object> recordSet) {
return formatCacheKey((String) recordSet.get("urn"));
}

@NotNull @Override
public CacheKey fromFieldArguments(@NotNull ResponseField field, @NotNull Operation.Variables variables) {
return formatCacheKey((String) field.resolveArgument("urn", variables));
}

private CacheKey formatCacheKey(String urn) {
if (urn == null || urn.isEmpty()) {
return CacheKey.NO_KEY;
} else {
return CacheKey.from(urn);
}
}
};

// There is supposed to be a sqlCache we can chain so we can read-write from the in-memory cache and allow trickle down writes
// to the sql cache, but in 1.0.0-alpha it didn't seem to exist
NormalizedCacheFactory cacheFactory = new LruNormalizedCacheFactory(EvictionPolicy.builder().maxSizeBytes(10 * 1024).build());

return ApolloClient
.builder()
.serverUrl(globalConfig.getGraphQlUrl())
// this works
//.httpCache(new ApolloHttpCache(cacheStore))
// this does not work
.normalizedCache(cacheFactory, resolver)
.okHttpClient(okHttpClient)
.build();
```
Environment:
Android Studio 3.2.1
Nexus 5X API 27 emulator

implementation 'com.apollographql.apollo:apollo-runtime:1.0.0-alpha'
implementation 'com.apollographql.apollo:apollo-rx2-support:1.0.0-alpha'
implementation 'com.apollographql.apollo:apollo-http-cache:1.0.0-alpha'

**Comment by digitalbuddha** [comment]
Howdy ho! https://github.com/apollographql/apollo-android/blob/master/apollo-android-support/src/main/java/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCacheFactory.java does in fact exist.  It is in the apollo-android artifact since we use the android version of sqlite.  Let me try to gather more info on the cache impl, my understanding was that it is consistent wtih other clients.

**Comment by sav007** [comment]
Seems that issue is related to the list of objects we don't call cache key resolver.

**Comment by bpollock-vida** [comment]
I don't understand.

> There is no reason to call cache key resolver for the child objects as cached records already embed all needed information how to lookup for the child objects.

Unless the cache key resolver is called, how can it be known how to properly lookup a child object?

2. On iOS if a cache miss happens with the key resolver it falls back to non-normalized resolving according to the query. I have observed that it will instead just miss cache and not return a value instead of falling back on android.

**Comment by sav007** [comment]
Pls see the dumped version of data stored in the store. 

> how can it be known how to properly lookup a child object by its normalized value?

It's stored already in store as CacheKeyReference with a reference to child object.

**Comment by bpollock-vida** [comment]
> Pls see the dumped version of data stored in the store.

Hey @sav007 thank you for the dump info. I don't think it is representative of the scenario I'm describing. In your example the root query returns **a single object** which resolves 'id'. In my case the root query returns **an array** which resolves to CacheKey.NO_KEY.

My issue is when there is a top level cache key miss.

Example:
```
apolloClient.query(query).httpCachePolicy(HttpCachePolicy.NETWORK_ONLY)
```

Issues:
1. Even on NETWORK_ONLY if caching fails, the Apollo Client is throwing an exception
2. On iOS the framework seems to fall back upon path-based caching if normalization fails, but we get an exception in #1
3.. The framework does not seem to keep looking for hits in children

You can probably simulate this by removing 'id' from Droid aka '2001' although returning an array containing Droid would probably be the best example.

### 1. Exception
I know this is caching, because everything works as expected if I comment out caching on the client configuration:
```
.normalizedCache(cacheFactory, resolver)
```

I have observed the following
1. fromFieldArguments calls formatCacheKey which returns CacheKey.NO_KEY
2. fromFieldRecordSet calls formatCacheKey which returns CacheKey.NO_KEY
3. Apollo returns an exception from the Apollo Client

```
W/System.err: com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response
        at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:101)
        at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor$1.onResponse(ApolloParseInterceptor.java:53)
        at com.apollographql.apollo.internal.interceptor.ApolloServerInterceptor$1$1.onResponse(ApolloServerInterceptor.java:94)
        at okhttp3.RealCall$AsyncCall.execute(RealCall.java:206)
        at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
    Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Object.toString()' on a null object reference
        at com.apollographql.apollo.api.ResponseField.asJsonValue(ResponseField.java:310)
        at com.apollographql.apollo.api.ResponseField.orderIndependentKey(ResponseField.java:296)
        at com.apollographql.apollo.api.ResponseField.cacheKey(ResponseField.java:240)
        at com.apollographql.apollo.internal.cache.normalized.ResponseNormalizer.willResolve(ResponseNormalizer.java:47)
        at com.apollographql.apollo.internal.response.RealResponseReader.willResolve(RealResponseReader.java:254)
        at com.apollographql.apollo.internal.response.RealResponseReader.readList(RealResponseReader.java:152)
        at <redacted>.<query>.map(<query>.java:354)
        <redacted>
        at com.apollographql.apollo.internal.response.RealResponseReader.readObject(RealResponseReader.java:138)
W/System.err:     at com.vida.client.tagging.GetTagsForGroupQuery$Data$Mapper.map(GetTagsForGroupQuery.java:239)
        at com.vida.client.tagging.GetTagsForGroupQuery$Data$Mapper.map(GetTagsForGroupQuery.java:234)
        at com.apollographql.apollo.response.OperationResponseParser$1.read(OperationResponseParser.java:96)
        at com.apollographql.apollo.internal.json.ResponseJsonStreamReader.nextObject(ResponseJsonStreamReader.java:86)
        at com.apollographql.apollo.response.OperationResponseParser.parse(OperationResponseParser.java:91)
        at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:87)
    	... 7 more
<redacted>: error: Stream Error com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response: com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response
```

### 2. Falling back on query caching
In [this Bug Report about cache normalization on iOS](https://github.com/apollographql/apollo-ios/issues/421) it shows we still get an array of tags as a response even if the cache resolves to NO_KEY. I have emulated this code on android and it throws an exception

### 3. The framework does not seem to keep looking for hits in children
Because it throws an exception, the client doesn't not try to normalize the children in this array response.

**Comment by JacquesSmuts** [comment]
> Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Object.toString()' on a null object reference
>         at com.apollographql.apollo.api.ResponseField.asJsonValue(ResponseField.java:310)
  
Are you sure this isn't happening because of this line:
```return formatCacheKey((String) recordSet.get("urn"));```

Which tries to cast a null to a String and throws an exception? What happens if you change that line to handle the null before trying to cast to String?

**Comment by bpollock-vida** [comment]
That doesn't seem to be the case @JacquesSmuts. I tried a bunch of stuff to be 'safer' but there is no exception from the string cast or my implementation. I upgraded from 1.0.0-alpha to alpha3 and have found:
1. Exception no longer happens
2. We do fall back on query caching and can get the collection as cached
3. normalization does not work

This is now identical to as described on iOS in:
https://github.com/apollographql/apollo-ios/issues/421
